pkg/apis/serving/v1alpha1/domainmapping_validation.go

```diff
@@ -20,43 +20,58 @@ import (
 	"context"
 	"fmt"
 
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/api/validation"
 	"knative.dev/pkg/apis"
-	duckv1 "knative.dev/pkg/apis/duck/v1"
+	"knative.dev/serving/pkg/apis/serving"
+	v1 "knative.dev/serving/pkg/apis/serving/v1"
 )
 
 // Validate makes sure that DomainMapping is properly configured.
-func (dm *DomainMapping) Validate(_ context.Context) *apis.FieldError {
-	return validateMetadata(&dm.ObjectMeta).ViaField("metadata").
-		Also(dm.validateSpec().ViaField("spec"))
-}
+func (dm *DomainMapping) Validate(ctx context.Context) *apis.FieldError {
+	errs := dm.validateMetadata(ctx).ViaField("metadata")
+
+	ctx = apis.WithinParent(ctx, dm.ObjectMeta)
+	errs = errs.Also(dm.Spec.Validate(apis.WithinSpec(ctx)).ViaField("spec"))
 
-// validateSpec validates the Spec of a DomainMapping.
-func (dm *DomainMapping) validateSpec() (errs *apis.FieldError) {
-	return errs.Also(validateRef(&dm.Spec.Ref, dm.Namespace).ViaField("ref"))
+	return errs
 }
 
-// validateRef validates the Ref field of a DomainMapping.
-func validateRef(ref *duckv1.KReference, ns string) (errs *apis.FieldError) {
-	if ref.Name == "" {
+// validateMetadata validates the metadata section of a DomainMapping.
+func (dm *DomainMapping) validateMetadata(ctx context.Context) (errs *apis.FieldError) {
+	if dm.Name == "" {
 		errs = errs.Also(apis.ErrMissingField("name"))
 	}
 
-	if ref.Namespace != "" && ref.Namespace != ns {
-		errs = errs.Also(&apis.FieldError{
-			Message: fmt.Sprintf("Ref namespace must be empty or equal to the domain mapping namespace %q", ns),
-			Paths:   []string{"namespace"},
-		})
+	if apis.IsInUpdate(ctx) {
+		original := apis.GetBaseline(ctx).(*DomainMapping)
+		errs = errs.Also(
+			apis.ValidateCreatorAndModifier(original.Spec, dm.Spec,
+				original.GetAnnotations(), dm.GetAnnotations(), serving.GroupName).ViaField("annotations"),
+		)
 	}
 
 	return errs
 }
 
-// validateMetadata validates the metadata section of a DomainMapping.
-func validateMetadata(md *metav1.ObjectMeta) (errs *apis.FieldError) {
-	if md.Name == "" {
-		return errs.Also(apis.ErrMissingField("name"))
+// Validate makes sure the DomainMappingSpec is properly configured.
+func (spec *DomainMappingSpec) Validate(ctx context.Context) *apis.FieldError {
+	errs := spec.Ref.Validate(ctx).ViaField("ref")
+
+	// For now, ref must be a serving.knative.dev/v1 Service.
+	if spec.Ref.Kind != "Service" {
+		errs = errs.Also(apis.ErrGeneric(`must be "Service"`, "ref.kind"))
```

Comment:
`Via("kind")` of sorts?

Reply:
think the final argument does this already

Reply:
Well `ref` is already part of the parent error. I'd just do errs.Also(errs.ViaField()...) or along those lines :)

Reply:
I think this is fine as-is. 🤷

Reply:
you mean like `errs = errs.Also(apis.ErrGeneric('must be "Service"', "kind").ViaField("ref"))`? Can do, seems to be a bit wordier without much benefit - or are you thinking of something else?

Reply:
Yeah, it's just that "ref.kind" is internal knowledge to how erros constructs field paths.
Tomorrow when it's "/" separated this will be a bug. Otherwise, it's fine.

Reply:
I'm not worried about a switch to XPath 😉